### PR TITLE
[JENKINS-45740] - Support injecting extra metadata into the Plugin manifest

### PR DESCRIPTION
**Pull request by oleg-nenashev**
This is a foundation work for [WEBSITE-383](https://issues.jenkins-ci.org/browse/WEBSITE-383)

- [x] - Inject extra metadata into the HPI Manifest file
  * `Plugin-ChangelogUrl` - custom changelog location, e.g. in GitHub
  * `Plugin-ScmUrl` - location of plugin sources
  * `Plugin-LogoUrl` - just a placeholder for the future. Some plugins like BlueOcean have their logos
  * `Plugin-Licence-Name${-N}` - `N` starts from 2, we need it for multi-license projects if they really exist.
  * `Plugin-Licence-Url-${N}`
- [x] - Integration tests
- [x] - Fix typos

All the data is being extracted from standard fields, excepting Changelog and Logo url. For this purpose extra properties should be set, no default values:

```xml
<properties>
        <jenkins.version>2.19.2</jenkins.version>
        <java.level>7</java.level>
        <plugin.info.changelogUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/changelog</plugin.info.changelogUrl>
        <plugin.info.logoUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png</plugin.info.logoUrl>
</properties>
```

https://issues.jenkins-ci.org/browse/JENKINS-45740

Reference implementation: https://github.com/jenkinsci/configuration-as-code-plugin/pull/989

@daniel-beck @reviewbybees

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Sample output:

```
Manifest-Version: 1.0
Archiver-Version: Plexus Archiver
Created-By: Apache Maven
Built-By: nenashev
Build-Jdk: 1.8.0_121
Extension-Name: filesystem_scm
Specification-Title: Using File System as SCM, done by checking file s
 ystem last modified date.
Implementation-Title: filesystem_scm
Implementation-Version: 1.21-alpha-2-SNAPSHOT
Group-Id: hudson.plugins.filesystem_scm
Short-Name: filesystem_scm
Long-Name: Jenkins File System SCM Plugin
Url: http://wiki.jenkins-ci.org/display/JENKINS/File+System+SCM
Plugin-Version: 1.21-alpha-2-SNAPSHOT (private-f5169a4c-nenashev)
Hudson-Version: 1.642.3
Jenkins-Version: 1.642.3
Plugin-Dependencies: structs:1.5
Plugin-Developers: Sam NG:samngms:samngms@yahoo.com,Christoph Kutzinsk
 i:kutzi:kutzi@gmx.de,Oleg Nenashev:oleg_nenashev:o.v.nenashev@gmail.com
Plugin-Licence-Name-1: MIT License
Plugin-Licence-Url-1: http://www.opensource.org/licenses/mit-license.php
Plugin-Licence-Distribution-1: repo
Plugin-ChangelogUrl: http://plugins.jenkins.io/Foo/
 changelog
Plugin-LogoUrl: http://plugins.jenkins.io/Foo/logo.
 png
Plugin-ScmUrl: http://github.com/jenkinsci/filesystem_scm-plugin
```

**Comment by oleg-nenashev** [comment]
This build downloads the world from Jenkins.IO. I will propose a patch for that

**Comment by daniel-beck** [comment]
I don't understand what the purpose of this is. The plugins site already extracts data from the POM and ignores metadata.mf as hard as possible; so why are these added there? Needs companion PR in the update center generator.

**Comment by oleg-nenashev** [comment]
> The plugins site already extracts data from the POM and ignores metadata.mf as hard as possible; so why are these added there? Needs companion PR in the update center generator.

Didn't know about that (should have checked it, yes). Then this change is not really required for WEBSITE-383, but it still may be useful for other cases.

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> Then this change is not really required for WEBSITE-383, but it still may be useful for other cases.

To clarify, there's no metadata corresponding to these changes, this is more about consistency in implementation. It would be good to have a design outline for this and the required changes to core, update center, plugin site, possibly maven-hpi-plugin.

File: src/it/JENKINS-45740-metadata/pom.xml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    <properties>
+        <hpi-plugin.version>@project.version@</hpi-plugin.version>
+        <plugin.info.changelogUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/changelog</plugin.info.changelogUrl>
+        <plugin.info.logoUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png</plugin.info.logoUrl>
+    </properties>
+    
+    <licenses>
+      <license>
+        <name>MIT License</name>
+        <url>http://www.opensource.org/licenses/mit-license.php</url>
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/pom.xml]: For OSS license URLs, please enforce a normal scheme: https://opensource.org/licenses/MIT

File: src/it/JENKINS-45740-metadata/src/main/java/test/SampleRootAction.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package test;
+
+import hudson.Extension;
+import hudson.model.InvisibleAction;
+import hudson.model.RootAction;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+
+@Extension
+public class SampleRootAction implements RootAction {
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/src/main/java/test/SampleRootAction.java]: delete

File: src/it/JENKINS-45740-metadata/src/test/java/test/SampleRootActionTest.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package test;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Test;
+import static org.junit.Assert.*;
+import org.junit.Rule;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class SampleRootActionTest {
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/src/test/java/test/SampleRootActionTest.java]: delete

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    'Plugin-Licence-Url-1': 'http://www.opensource.org/licenses/mit-license.php',
+    'Plugin-Licence-Distribution-1': 'repo',
+    'Plugin-Licence-Comments-1': null,
+    'Plugin-Licence-Name-2': 'MY License',
+    'Plugin-Licence-Url-2': 'http://mylicense.txt',
+    'Plugin-Licence-Distribution-2': null,
+    'Plugin-Licence-Comments-2': "Hello, world!"
+    ])
+
+// TODO try mvn hpi:run (how do we find the right version of mvn to run?) and check that we can access http://localhost:8080/jenkins/sample/
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: delete all this copy-pasted stuff irrelevant for this IT

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+// TODO run a copy of jenkins.war with the installed *.hpi predeployed and do a similar check
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: delete

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+return true
+
+// TODO make similar test for plugin with no-test-jar=true (the default)
+// TODO also test building with: JITPACK=true mvn -DskipTests clean install
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: delete

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+File installed = new File(basedir, '../../local-repo/org/jenkins-ci/tools/hpi/its/JENKINS-45740-metadata/1.0-SNAPSHOT/')
+assert installed.directory
+
+def checkArtifact(File installed, String artifact, List<String> expectedEntries, List<String> unexpectedEntries, Map<String,String> expectedManifestEntries) {
+    File f = new File(installed, artifact)
+    assert f.file
+    JarFile j = new JarFile(f)
+    try {
+        expectedEntries.each {entry -> assert j.getEntry(entry) != null : "no ${entry} in ${f}"}
+        unexpectedEntries.each {entry -> assert j.getEntry(entry) == null : "unwanted ${entry} in ${f}"}
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: delete

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+}
+
+checkArtifact(installed, 'JENKINS-45740-metadata-1.0-SNAPSHOT.hpi',
+    [],[],
+    ['Plugin-ChangelogUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/changelog',
+    'Plugin-LogoUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png',
+    'Plugin-ScmUrl': 'http://github.com/jenkinsci/JENKINS-45740-metadata-plugin',
+    'Plugin-Licence-Name-1': 'MIT License',
+    'Plugin-Licence-Url-1': 'http://www.opensource.org/licenses/mit-license.php',
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: see above

File: src/it/JENKINS-45740-metadata/pom.xml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        <distribution>repo</distribution>
+      </license>
+      <license>
+        <name>MY License</name>
+        <url>http://mylicense.txt</url>
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/pom.xml]: How can we supporting two licenses? This seems like a nonexistent use case.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 10]
```diff
@@ -1021,6 +1046,28 @@ private String getDevelopersForManifest() throws IOException {
         return buf.toString();
     }
 
+    private void addLicenseAttributesForManifest(Section target) throws ManifestException {
+        final List licenses = project.getLicenses();
+        int licenseCounter = 1;
+        for (Object obj : licenses) {
+            if (obj instanceof License) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: A safe cast, no need to check it.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java [hunk 11]
```diff
@@ -1021,6 +1046,28 @@ private String getDevelopersForManifest() throws IOException {
...
+                License lic = (License)obj;
+                addAttributeIfNotNull(target, "Plugin-Licence-Name-" + licenseCounter, lic.getName());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java]: British/etc. variant, potentially confusing given the mismatch with the American variant adopted by Maven.

[Background](http://grammarist.com/spelling/licence-license/)

File: src/it/JENKINS-45740-metadata/pom.xml (hunk 9 above)
Comment (jglick) [inline src/it/JENKINS-45740-metadata/pom.xml]: Do we actually have any such plugins?

File: src/it/JENKINS-45740-metadata/pom.xml (hunk 1 above)
Comment (oleg-nenashev) [inline src/it/JENKINS-45740-metadata/pom.xml]: agreed

Comment (oleg-nenashev) [reply src/it/JENKINS-45740-metadata/pom.xml]:
```suggestion
        <url>https://opensource.org/licenses/MIT</url>
```

File: src/it/JENKINS-45740-metadata/pom.xml (hunk 9 above)
Comment (oleg-nenashev) [reply src/it/JENKINS-45740-metadata/pom.xml]: AFAIK there is no dual-licensed plugins. I do not recall why I did that, but we could revert it to a single-licence mode to K.I.S.S.

**Comment by oleg-nenashev** [comment]
the work is not really in progress, but I would be willing to finish it if there is interest from the community

**Comment by halkeye** [comment]
I'm interested in it. I would like to get GitHub release notes onto the plugin site.

**Comment by oleg-nenashev** [comment]
The code might be still operational, actually. I merged the PR with master, let;s see what is left to be done here. From what I see, themost of the comments from @jglick are about the integration test cleanup. 

We will also need some docs to be added

**Comment by halkeye** [comment]
Looks like some variables got added/deleted? i will dig into it when i get a chance

```
[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[971,13] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   variable compatibleSinceVersion

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[972,88] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   variable compatibleSinceVersion

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[974,13] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   variable sandboxStatus

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[975,78] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   variable sandboxStatus

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[981,74] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   class Date

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[981,26] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   class SimpleDateFormat

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo

[2019-06-27T22:32:42.066Z] [ERROR] /home/jenkins/workspace/Plugins_maven-hpi-plugin_PR-68/src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java:[1127,28] cannot find symbol

[2019-06-27T22:32:42.066Z]   symbol:   class Collection

[2019-06-27T22:32:42.066Z]   location: class org.jenkinsci.maven.plugins.hpi.AbstractHpiMojo
```

**Comment by halkeye** [comment]
Also side note:
```
[2019-06-27T22:32:33.935Z] [ERROR] 

[2019-06-27T22:32:33.935Z] 

[2019-06-27T22:32:33.935Z] Artifact Ids of the format maven-___-plugin are reserved for 

[2019-06-27T22:32:33.935Z] plugins in the Group Id org.apache.maven.plugins

[2019-06-27T22:32:33.935Z] Please change your artifactId to the format ___-maven-plugin

[2019-06-27T22:32:33.935Z] In the future this error will break the build.
```

**Comment by halkeye** [comment]
@oleg-nenashev do you want me to do the cleanup to address jesse's comments?

**Comment by oleg-nenashev** [comment]
Feel free to do so. I have not yet got to them, and I do not want to block
you

**Comment by halkeye** [comment]
I'm not blocked per say. I got a billion things to do and rotate between them. Next pass I'll delete the stuff

**Comment by halkeye** [comment]
[2019-07-07T05:29:55.883Z] [WARNING] Could not validate integrity of download from https://repo.jenkins-ci.org/public/org/kohsuke/stapler/stapler-parent/1.257-JEP-216-SNAPSHOT/stapler-parent-1.257-JEP-216-20190205.205818-1.pom.sha1: Checksum validation failed, no checksums available


I think it just needs a retry, but I'll poke more later

**Comment by halkeye** [comment]
More updates for oleg on his repo! https://github.com/oleg-nenashev/maven-hpi-plugin/pull/23

**Comment by jglick** [comment]
Do I understand correctly that the immediate motivation here is being able to declare a `changelogUrl` to point to GitHub Releases that could then be consumed by plugins.jenkins.io and `/pluginManager`/? If so, there will need to be some downstream PRs demonstrating consumption of the new metadata.

**Comment by halkeye** [comment]
Yea.

Last night i got it running inside a debugger and saw that it's not doing as expected due to how old it is and all the new code.

So this pr needs a decent amount of work too

**Comment by halkeye** [comment]
https://github.com/oleg-nenashev/maven-hpi-plugin/pull/24 should fix the PR so the tests run again

Next up for me is to figure out how to get the plugin site backend to pull a local compiled pom to test that out. Or maybe just a WIP PR for plugins site with hardcoded data

**Comment by oleg-nenashev** [comment]
@halkeye to save your time, you could just start a new pull request on the top of the current one. In such case you won't need to propose pull requests in my local fork

**Comment by halkeye** [comment]
Yep but that should be all needed changes.

Should be blue now

All other changes will be in other repos

**Comment by halkeye** [comment]
Yay check mark

File: src/it/JENKINS-45740-metadata/pom.xml [hunk 12]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>plugin</artifactId>
+        <version>3.29</version>
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/pom.xml]:
```suggestion
        <version>3.47</version>
```

File: src/it/JENKINS-45740-metadata/pom.xml (hunk 9 above)
Comment (jglick) [reply src/it/JENKINS-45740-metadata/pom.xml]: Yes I would suggest removing something we do not think we actually need, unless and until a clear need arises.

File: src/it/JENKINS-45740-metadata/src/test/resources/test/expected.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+sample served
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/src/test/resources/test/expected.txt]: Unused AFAICT, could be deleted.

File: src/it/JENKINS-45740-metadata/pom.xml [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    </parent>
+    <groupId>org.jenkins-ci.tools.hpi.its</groupId>
+    <artifactId>JENKINS-45740-metadata</artifactId>
+    <version>1.0-SNAPSHOT</version>
+    <packaging>hpi</packaging>
+    <properties>
+        <java.level>8</java.level>
+        <hpi-plugin.version>@project.version@</hpi-plugin.version>
+        <plugin.info.changelogUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/changelog</plugin.info.changelogUrl>
+        <plugin.info.logoUrl>http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png</plugin.info.logoUrl>
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/pom.xml]: Are these realistic examples? I would have thought the changelog URL would be `…/releases` and a logo URL would be a GitHub raw content link.

I guess for purposes of the test they could be any well-formed URLs?

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+checkArtifact(
+    installed,
+    'JENKINS-45740-metadata-1.0-SNAPSHOT.hpi',
+    [
+      'Plugin-ChangelogUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/changelog',
+      'Plugin-LogoUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png',
+      'Plugin-ScmUrl': 'http://github.com/jenkinsci/JENKINS-45740-metadata-plugin',
+      'Plugin-License-Name-1': 'MIT License',
+      'Plugin-License-Url-1': 'https://opensource.org/licenses/MIT',
+      'Plugin-License-Distribution-1': 'repo',
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: I have never even understood what this field of the POM is for, but I think we do not need it. Really I would suggest dropping the name too, and only support _canonical_ URLs: `https://opensource.org/licenses/.+`

File: src/it/JENKINS-45740-metadata/verify.groovy [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    }
+}
+
+checkArtifact(
+    installed,
+    'JENKINS-45740-metadata-1.0-SNAPSHOT.hpi',
+    [
+      'Plugin-ChangelogUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/changelog',
+      'Plugin-LogoUrl': 'http://plugins.jenkins.io/JENKINS-45740-metadata/logo.png',
+      'Plugin-ScmUrl': 'http://github.com/jenkinsci/JENKINS-45740-metadata-plugin',
```
Comment (jglick) [inline src/it/JENKINS-45740-metadata/verify.groovy]: Let us require HTTPS.

File: src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractJenkinsManifestMojo.java [hunk 17]
```diff
@@ -266,4 +275,39 @@ protected Manifest loadManifest(File f) throws IOException, ManifestException {
...
     }
+
+    private void addLicenseAttributesForManifest(Manifest.Section target) throws ManifestException {
+        final List<License> licenses = project.getLicenses();
+        int licenseCounter = 1;
+        for (License lic : licenses) {
+            addAttributeIfNotNull(target, "Plugin-License-Name-" + licenseCounter, lic.getName());
+            addAttributeIfNotNull(target, "Plugin-License-Url-" + licenseCounter, lic.getUrl());
+            addAttributeIfNotNull(target, "Plugin-License-Distribution-" + licenseCounter, lic.getDistribution());
+            addAttributeIfNotNull(target, "Plugin-License-Comments-" + licenseCounter, lic.getComments());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractJenkinsManifestMojo.java]: Comments? Really? Let us only support the attributes that we have actual, proven use cases for in downstream PRs.

**Comment by oleg-nenashev** [comment]
Does it really worth rewriting the implementation @jglick ? Some fields might be YAGNI, but they do not hurt. Same for the license, it does not matter much if we dump multiple licenses (though we could remove index for the first one). I am just a bit lazy to clean it up, yeah

**Comment by jglick** [comment]
> Some fields might be YAGNI, but they do not hurt.

Are we ready to delete unused manifest attributes later?

**Comment by oleg-nenashev** [comment]
@halkeye @jglick @jenkinsci/core 

I suggest the following plan here:

* We remove `Plugin-License-Distribution-X` and `Plugin-License-Comment-X` fields. I see no harm in keeping them, but I am happy to remove them to unblock the PR
* Tests will be updated to use HTTPs
* Multi-license support will be kept. If we ever start REALLY bothering about being compliant with MIT and Apache licenses, we will need that. Many plugins de-facto include both MIT and Apache v.2 code

If we agree on such plan, I will get it implemented.

**Comment by jglick** [comment]
> to unblock the PR

I am not blocking this PR, just asking some questions and leaving it to your & @daniel-beck’s judgment how to proceed. From my PoV the main question is whether the downstream PRs have been prepared and are demonstrably able to consume this new metadata in order to add some user-visible features.

File: src/it/JENKINS-45740-metadata/pom.xml (hunk 14 above)
Comment (oleg-nenashev) [reply src/it/JENKINS-45740-metadata/pom.xml]: JCasC, BlueOcean, unofficial Pipeline logos.
Also, many tool/service integration plugins can star using tool logos

Comment (halkeye) [reply src/it/JENKINS-45740-metadata/pom.xml]: If the changelog url points to the plugin site, how will the plugin site know where to get changelogs?

Comment (halkeye) [reply src/it/JENKINS-45740-metadata/pom.xml]: Like one value would be for update center and one plugins site? Or would update center just always assume plugin site

Comment (oleg-nenashev) [reply src/it/JENKINS-45740-metadata/pom.xml]: This is just a sample URL, does not really matter what;s the content. In real use-cases I expect it to link to GitHub Releases or other similar sources. Probably it is a good time to start writing documentation for the plugin...

**Comment by oleg-nenashev** [comment]
Ready to go?

**Comment by jglick** [comment]
Are there downstream/linked PRs yet? If so, please list them in this PR’s description.

**Comment by oleg-nenashev** [comment]
No downstream PRs so far. I can create one in JCasC if it helps, but basically it is what's written in the integration test. https://github.com/jenkins-infra/plugin-site-api will be the main consumer

**Comment by oleg-nenashev** [comment]
Deployed it as `3.7-20190820.090442-1`

**Comment by oleg-nenashev** [comment]
Added reference implementation: https://github.com/jenkinsci/configuration-as-code-plugin/pull/989

**Comment by oleg-nenashev** [comment]
@halkeye @jglick gentle ping, I think it is ready to go. Once merged, I will release several plugins to have initial metadata for the Plugin Site

**Comment by jglick** [comment]
So we are still awaiting usage in `plugin-site-api`? Did this also need changes in `jenkins` to display metadata in **Plugin Manager**?

**Comment by oleg-nenashev** [comment]
> So we are still awaiting usage in plugin-site-api?

Nope, we need data for it first. 

>  Did this also need changes in jenkins to display metadata in Plugin Manager?

Maybe later. I do see benefits in displaying logos and changelogs straight inside the Update Center, but right now I rather focus https://issues.jenkins-ci.org/browse/WEBSITE-637

**Comment by zbynek** [comment]
@oleg-nenashev do we also need changes in update-center2 for this to work?  Plugin site gets the metadata from there AFIK.

**Comment by oleg-nenashev** [comment]
Yes, we will need it AFAIK. That's why I want to have an https://issues.jenkins-ci.org/browse/WEBSITE-637 EPIc to list all stories

**Comment by daniel-beck** [comment]
> logoUrl

Could we replace this with a base64 image definition so we don't make everyone call to maintainer specified random URLs whose content can change over time or disappear?

**Comment by daniel-beck** [comment]
> > logoUrl
> 
> Could we replace this with e.g. a base64 image definition so we don't make everyone call to maintainer specified random URLs whose content can change over time or disappear, and that breaks in offline scenarios anyway?

I guess with `data:` scheme this is compatible; can probably just reject any other scheme in consumers.

**Comment by oleg-nenashev** [comment]
> Could we replace this with e.g. a base64 image definition so we don't make everyone call to maintainer specified random URLs whose content can change over time or disappear, and that breaks in offline scenarios anyway?

Could be an additional field or format. Definitely fine with me

**Comment by zbynek** [comment]
Couldn't the POM contain location of an image file inside the repo and then it would get encoded as base64 in the manifest?

**Comment by daniel-beck** [comment]
@zbynek That's a great idea!

**Pull request merged**